realm/realm-library/src/androidTest/java/io/realm/DynamicRealmTests.java

```diff
@@ -996,4 +999,187 @@ public void realmListRemoveAllFromRealm() {
         assertEquals(0, list.size());
         assertEquals(0, realm.where(Dog.CLASS_NAME).count());
     }
+
+    // Test if waitForChange gets waked up by 1) empty transaction 2) transaction with some data changes.
+    @Test
+    public void waitForChange() throws InterruptedException {
+        final long numberOfBlocks = 16;
+        final long numberOfObjects = 1; // must be greater than 1
+        final CountDownLatch openedLatch = new CountDownLatch(1);
+        final CountDownLatch changedLatch = new CountDownLatch(1);
+        final CountDownLatch closedLatch = new CountDownLatch(1);
+        final AtomicBoolean result1 = new AtomicBoolean(false);
+        final AtomicBoolean result2 = new AtomicBoolean(false);
+        final AtomicLong count1 = new AtomicLong(0);
+        final AtomicLong count2 = new AtomicLong(0);
+
+        // wait in background
+        final CountDownLatch signalTestFinished = new CountDownLatch(1);
+        Thread thread = new Thread(new Runnable() {
+            @Override
+            public void run() {
+                DynamicRealm realm = DynamicRealm.getInstance(defaultConfig);
+                openedLatch.countDown();
+                result1.set(realm.waitForChange());
+                count1.set(realm.where(AllTypes.CLASS_NAME).count());
+                changedLatch.countDown();
+                result2.set(realm.waitForChange());
+                count2.set(realm.where(AnnotationIndexTypes.CLASS_NAME).count());
+                realm.close();
+                closedLatch.countDown();
+            }
+        });
+        thread.start();
+
+        TestHelper.awaitOrFail(openedLatch);
+        Thread.sleep(100);
```

Comment:
We should never have sleep in our tests. It indicates a race condition you don't fully control (so will result in flaky tests).